sources/NSStringITerm.m

```diff
@@ -2196,11 +2196,34 @@ - (void)escapeShellCharactersExceptTabAndNewline {
 }
 
 - (void)escapeShellCharactersIncludingNewlines:(BOOL)includingNewlines {
-    NSString* charsToEscape = [NSString shellEscapableCharacters];
-    if (includingNewlines) {
-        charsToEscape = [charsToEscape stringByAppendingString:@"\r\n"];
+    if ([iTermAdvancedSettingsModel escapeWithQuotes]) {
+        // Only need to escape single quote and backslash in a single-quoted string
+        NSString* charsToEscape = @"\\'";
+        NSString* charsToSearch = [NSString shellEscapableCharacters];
```

Comment:
Instead of creating a string and then iterating over its characters, you can build a NSMutableCharacterSet and insert the characters you want to search for. Then use `-[NSString rangeOfCharacterFromSet:]` to see if any exists.

Iterating over characters in a string has a lot of ways it can fail (surrogate pairs and combining marks, for example) and is also slower than using a character set.